Created Fido Client for Gong synoptic maps

abhijeetmanhas opened the pull request:
<!--
We know that working on sunpy and submitting pull requests takes effort, and we appreciate your time.
Thank you.

Please be aware that everyone has to follow our code of conduct:
https://github.com/sunpy/sunpy/blob/master/CODE_OF_CONDUCT.rst

Also these comments are hidden when you submit this github pull request.

We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry about them!
Here is a brief explanation of them: https://docs.sunpy.org/en/latest/dev_guide/pr_review_procedure.html#continuous-integration.
-->

### Description
<!--
Provide a general description of what your pull request does.

If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number GitHub will automatically link it.
If it doesn't, please remove the following line.
-->

Fixes https://github.com/dstansby/pfsspy/issues/165

This will enable users to search and download the meged data product for GONG using Fido, giving the timerange. The url pattern is like this,

`https://gong2.nso.edu/oQR/zqs/%Y%m/mrzqs%y%m%d/mrzqs%y%m%dt%H%Mc\d4_\d3.fits.gz`
Here `\d4` and `\d3` are Carringtion rotation number and Initital carr-long for synoptic map respectively.

Used scraper partially, I can't write client only using scraper since it scraper needs everything to be defined in the pattern, which won't be the case for these two attributes. It doesn't supports custom regex.

abhijeetmanhas commented:
Main things to bee reviewed:

- The name of Client
- The `_makemap` routine, what more can be added ?
- Should add its own fetch function, which will extract the files too?
- Add a function to convert the fits data to pfsspy input compatible ?
- Should allow users to specify the `CAR_ROT` and `LONG0` ?
- `_can_handle_query` , what more `phsyobs`, `level`, etc should it support ?

Sunpy scraper should be able to handle custom regex, something by which we can solve issue https://github.com/sunpy/sunpy/issues/3337 by making scraper selecting only those goes_satellites for which data is present. Can be handled in different PR.

pep8speaks commented:
Hello @abhijeetmanhas! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`sunpy/net/dataretriever/sources/gong_synoptic.py`](https://github.com/sunpy/sunpy/blob/8835f80687f82e1c588150a539dfeabda3670e65/sunpy/net/dataretriever/sources/gong_synoptic.py):

> [Line 56:30](https://github.com/sunpy/sunpy/blob/8835f80687f82e1c588150a539dfeabda3670e65/sunpy/net/dataretriever/sources/gong_synoptic.py#L56): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\d'
> [Line 56:36](https://github.com/sunpy/sunpy/blob/8835f80687f82e1c588150a539dfeabda3670e65/sunpy/net/dataretriever/sources/gong_synoptic.py#L56): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\d'
> [Line 56:42](https://github.com/sunpy/sunpy/blob/8835f80687f82e1c588150a539dfeabda3670e65/sunpy/net/dataretriever/sources/gong_synoptic.py#L56): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\d'
> [Line 56:46](https://github.com/sunpy/sunpy/blob/8835f80687f82e1c588150a539dfeabda3670e65/sunpy/net/dataretriever/sources/gong_synoptic.py#L56): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\d'

dstansby commented:
> Main things to bee reviewed:
> 
>     * The name of Client
Seems fine to me
> 
>     * The `_makemap` routine, what more can be added ?
> 
I'm not sure what this is, so pass...
>     * Should add its own fetch function, which will extract the files too?
> 
Do you mean un-zip them from `.fits.gz` to `.fits`?
>     * Add a function to convert the fits data to pfsspy input compatible ?
> 
That's out of scope for sunpy, so no
>     * Should allow users to specify the `CAR_ROT` and `LONG0` ?
> 
I think searching by time is fine
>     * `_can_handle_query` , what more `phsyobs`, `level`, etc should it support ?
> 
Again, just searching by time is fine. We can always expand this later.
> 
> Sunpy scraper should be able to handle custom regex, something by which we can solve issue #3337 by making scraper selecting only those goes_satellites for which data is present. but it can be handled in different PR.

abhijeetmanhas commented:
Yes, I was thinking of unzipping them into fits. Should we do it here? Or just leave gz files downloaded?

dstansby commented:
Ideally we want to unzip them, if possible.

Cadair commented:
I am pretty sure the SRS client handles unpacking if you want some prior art.

abhijeetmanhas commented:
@Cadair thanks for the suggestion.
I discovered your comment after pushing. SRS uses Tarfile library, I am using gzip. 
Which one would be better here?

abhijeetmanhas commented:
@dstansby the test failure is remote server error, rest all tests have passed. It is good for review now.

dstansby commented on sunpy/net/dataretriever/sources/gong_synoptic.py:
```diff
@@ -0,0 +1,157 @@
+        boolean
+            answer as to whether client can service the query
+        """
+        chkattr = ['Time', 'Instrument']
+        chklist = [x.__class__.__name__ in chkattr for x in query]
+        for x in query:
+            if x.__class__.__name__ == 'Instrument' and x.value.lower() == 'gong':
+                return all(chklist)
+        return False
```
It is possible to download different types of data from GONG, so I think we need to put some more thought into making `_can_handle_query` more specific. I think what we might need is an attribute that specifies we are looking for a synoptic map (as opposed to full disk observation), and an attribute that specifies that we are looking for magnetic field (more specifically, radial magnetic field) data. I think I will open this in a new issue, so people can see/comment on ideas for these attributes (and whether we need to make any new ones).

abhijeetmanhas replied:
Thanks for the reivew, I have added physobs and extentType now.

abhijeetmanhas replied:
@dstansby this is green now, can be reviewed, I have done all suggested changes till now.

dstansby commented:
Thanks for all the work on this! On the technical side it looks 👍. As I've said in a comment, more thought needs to go in on the sunpy/Fido API side to come up with a more specific way to search for this data. I will open a new issue to discuss that.

nabobalis commented:
Would a specific attr help?

abhijeetmanhas commented:
@nabobalis I have added optional `ExtentType` attrs like satellite no. in goes to differentiate b/w `FULL_DISK` and `SYNOPTIC` data, and tests for the same too. Also I have added optional `physobs` which should by `LOS_MAGNETIC_FIELD` as discussed with @dstansby .

dpshelio commented on sunpy/net/dataretriever/sources/gong_synoptic.py:
```diff
@@ -0,0 +1,172 @@
+        for directory in directories:
+            try:
+                opn = urlopen(directory)
+                try:
+                    soup = BeautifulSoup(opn, "html.parser")
+                    for link in soup.find_all("a"):
+                        href = link.get("href")
+                        if re.match(file_pattern, href):
+                            # extracting timestamp from the filename
+                            timestring = href[5:16]
+                            tstamp = parse_time('20'+timestring.upper()+'00')
+                            if tstamp >= timerange.start and tstamp <= timerange.end:
+                                fullpath = directory + href
+                                filesurls.append(fullpath)
```
Shouldn't all that be "done" by the scrapper? I believe you can put a wildcard if you don't know how it goes (like the last digits)... Ok, you can't! but with my new pr #4107

abhijeetmanhas replied:
This PR will be very helpful. I'll change the code to depend on #4107 after it is merged.

dpshelio commented on sunpy/net/dataretriever/sources/gong_synoptic.py:
```diff
@@ -0,0 +1,172 @@
+
+        resobj = downloader.download()
+
+        for loc_fpath, fpath in zip(local_paths, paths):
+            f = gzip.open(fpath, 'r')
+            g = gzip.open(loc_fpath, 'w')
+            g.write(f.read())
```
Why is this needed?

dpshelio replied:
Aren't you removing the `.gz` on line 112? why are you then zipping it back without the `gz` suffix?

abhijeetmanhas replied:
Yes, I removed `.gz` in line 112 to just get what the filename of extracted file should be. I haven't done any extraction there.
`local_fpath` contains path to future extracted file, which is currently empty when I opened it in write mode, and write `f` object to it, which contains the extracted file object in `fits`.

dpshelio replied:
sorry I've not commented earlier. I don't understand why you want to unzip the file. I believe fits, and map, reads zipped files fine. I'd prefer to keep the files zipped in my computer.

abhijeetmanhas replied:
Cool, keeping it zipped then. Thanks for the review!

nabobalis commented on changelog/4055.feature.rst:
```diff
@@ -0,0 +1,2 @@
+Created `~sunpy.net.dataretriever.sources.GongSynopticClient` for accessing merged data product
```
What is a merged data product?

abhijeetmanhas replied:
Changed to 
```Created `~sunpy.net.dataretriever.sources.GongSynopticClient` for accessing Magnetogram Synoptic Map Archives of NSO-GONG.```.

abhijeetmanhas commented:
`helio` Test failure is not related to the change.

dpshelio commented on sunpy/net/dataretriever/sources/tests/test_gong_synoptic.py:
```diff
@@ -0,0 +1,112 @@
+    assert isinstance(qr, UnifiedResponse)
+    response = Fido.fetch(qr)
+    assert len(response) == qr._numfile
+    for res in response:
+        assert res.split('.')[-1] == 'gz'
```
I don't really like `assert` on loops, if one fails, you only get a "report" of that one, and not whether others did too.
```suggestion
    assert all(map(lambda x: x.endswith('.gz'), response))
```

nabobalis commented on sunpy/net/dataretriever/sources/gong_synoptic.py:
```diff
@@ -0,0 +1,105 @@
+    """
+    Provides access to the Magnetogram products of NSO-GONG synoptic Maps.
+
+    Searches data hosted by the `National Solar Observatory <gong2.nso.edu/oQR/zqs/>`__
```
```suggestion
    Searches data hosted by the `National Solar Observatory. <gong2.nso.edu/oQR/zqs/>`__
```

nabobalis commented on sunpy/net/dataretriever/sources/gong_synoptic.py:
```diff
@@ -0,0 +1,105 @@
+
+    Examples
+    --------
+
```
```suggestion
```

nabobalis commented on sunpy/net/dataretriever/sources/gong_synoptic.py:
```diff
@@ -0,0 +1,105 @@
+        Parameters
+        ----------
+        timerange : `~sunpy.time.TimeRange`
+            The time range you want the files for.
```
```suggestion
            The time range required.
```

nabobalis commented on sunpy/net/dataretriever/sources/gong_synoptic.py:
```diff
@@ -0,0 +1,105 @@
+
+        Returns
+        -------
+        boolean
```
```suggestion
        bool : 
```

nabobalis commented on sunpy/net/dataretriever/sources/gong_synoptic.py:
```diff
@@ -0,0 +1,105 @@
+        Returns
+        -------
+        boolean
+            answer as to whether client can service the query
```
```suggestion
            Answer as to whether client can service the query,
```

nabobalis commented on sunpy/net/dataretriever/sources/gong_synoptic.py:
```diff
@@ -0,0 +1,105 @@
+        boolean
+            answer as to whether client can service the query
+        """
+        from sunpy.net import attrs as a
```
Does this need to be in the function?

abhijeetmanhas replied:
I thought it prevents circular imports.

nabobalis replied:
Can you check?

abhijeetmanhas replied:
yes, it prevents. https://github.com/sunpy/sunpy/blob/master/sunpy/net/dataretriever/sources/goes.py#L380 similar to what we did for `goes` and `norh`.

Cadair commented on sunpy/net/dataretriever/sources/gong_synoptic.py:
```diff
@@ -0,0 +1,103 @@
+                matches = False
+            if isinstance(x, a.Physobs) and x.value.lower() != 'los_magnetic_field':
+                matches = False
+            if isinstance(x, a.gong_synoptic.ExtentType) and x.value.lower() != 'synoptic':
```
Why do we have an attr for a thing we only support one type of?

dstansby replied:
I was thinking that we could re-use the VSO `EXTENT_TYPE` keyword here, but it's currently broken: https://github.com/sunpy/sunpy/issues/4048

Cadair replied:
If we don't have any use for this attr I suggest we don't implement it at this time. The extent type for this new client should be set in the `_map` metadata in the client class.

abhijeetmanhas replied:
I added extentType after this https://github.com/sunpy/sunpy/pull/4055#discussion_r414496406 . We can either remove it too.

Cadair replied:
Is gong data with different extent types available through the VSO? but we can't filter on it because of #4048 ?

dstansby replied:
No, I don't think GONG synoptic maps are on the VSO, but I think it would be good if we can re-purpose the EXTENT_TYPE attr in VSO for search (ie. move it from a vso specific attr to a general one like `Time` or `Instrument`) to avoid creating new attrs.

dstansby replied:
Okay, I did some digging/research and the EXTENT_TYPE attr for VSO is properly broken, so I think including a new attr is fine here.

If it was working then these are the valid values: https://sdac.virtualsolar.org/cgi/show_details?keyword=EXTENT_TYPE - @Cadair do you think it's worth putting the new attr in `sunpy.net.attrs` instead of in `sunpy.net.gong_synoptic.attrs`?

dstansby replied:
Going to resolve this conversation, and move discussion on where this new attr should go to an issue.

dstansby commented on sunpy/net/dataretriever/sources/gong_synoptic.py:
```diff
@@ -0,0 +1,103 @@
+    3 Results from the GongSynopticClient:
+         Start Time           End Time      Source Instrument Wavelength
+    ------------------- ------------------- ------ ---------- ----------
+    2019-12-31 22:00:00 2020-01-01 02:00:00    NSO       GONG        nan
+    2019-12-31 22:00:00 2020-01-01 02:00:00    NSO       GONG        nan
+    2019-12-31 22:00:00 2020-01-01 02:00:00    NSO       GONG        nan
```
This result looks odd, do you know why this is picking up three files, but all with the same timestamp?

abhijeetmanhas replied:
Fixed now.

dstansby commented:
@nabobalis which PR is this waiting on?

nabobalis commented:
> @nabobalis which PR is this waiting on?

https://github.com/sunpy/sunpy/pull/4321

Cadair commented:
#4321 has been merged so this also needs a rewrite.

nabobalis commented on sunpy/net/dataretriever/sources/gong_synoptic.py:
```diff
@@ -0,0 +1,46 @@
+        adict = {attrs.Instrument: [("GONG", "Global Oscillation Network Group.")],
+                 attrs.Physobs: [("LOS_MAGNETIC_FIELD", "Line of sight magnetic field")],
+                 attrs.Source: [('NSO', 'National Solar Observatory.')],
+                 attrs.gong_synoptic.ExtentType: [("SYNOPTIC", "Coverage of a complete solar rotation synthesized over time")]}
```
I don't have a better suggestion but `attrs.gong_synoptic.ExtentType` seems a bit long.

abhijeetmanhas replied:
Can we just remove it?
It will only happen that there won't be an extra column of name `extent type` in results whoes value is "synoptic". 
Nothing else would change.

dstansby replied:
I think it would be reasonable to move `ExtentType` out of `attrs.gong_synoptic` and into `attrs` as a more generic attribute, don't know what others think about that?

nabobalis replied:
IF we need to use it for this client, I would be in favour of that instead of it being for this client only.

dstansby replied:
Given one can also get full disc magnetograms from GONG, I actually think we should rename the client from `gong_synoptic` > `gong`, and keep this keyword as a way to differentiate between a full disk map and a synoptic map. Then we avoid having separate `gong` and a `gong_synoptic` clients.

abhijeetmanhas replied:
Replaced gongsynoptic with gong and moved extentType to attrs.

nabobalis commented on sunpy/net/attrs.py:
```diff
@@ -20,10 +20,12 @@
 * `a.vso <sunpy.net.vso.attrs>`
 * `a.jsoc <sunpy.net.jsoc.attrs>`
 * `a.goes <sunpy.net.dataretriever.attrs.goes>`
+* `a.gong_synoptic <sunpy.net.dataretriever.attrs.gong_synoptic>`
```
```suggestion
```
